tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php

```diff
@@ -271,7 +272,7 @@ public function testGeneratesConstraintCreationSql() : void
 
     public function getGenerateForeignKeySql() : string
     {
-        return null;
+        throw new LogicException('Foreign key constraints are not yet supported for SQLite.');
```

Comment:
> This is not caught by tests as this method is never called because `SqlitePlatform` does not support exceptions.

I'm curious why this is not flagged by PHPStan. I'm not a fan of fixing issues like this one by one. There may be more issues like this including the production code.

As for the code change, I'd rather fail the test explicitly with `$this->fail()`.

Reply:
> I'm curious why this is not flagged by PHPStan.

I've seen your work related to phpstan earlier. Did you run it on `tests`, not just `lib`?

> As for the code change, I'd rather fail the test explicitly with $this->fail().

I'll happily change it if you think it's better, but be advised that it makes it harder to understand for static analysis, as the method does not return or throw anything. The throw is hidden in `fail()`.

Reply:
> I've seen your work related to phpstan earlier. Did you run it on `tests`, not just `lib`?

It should run on `lib` and `tests`:

https://github.com/doctrine/dbal/blob/644531f360e48d4f222ccfd726a07aed60200b3e/phpstan.neon.dist#L4-L5

> […] it makes it harder to understand for static analysis, as the method does not return or throw anything.

AFAIK the PHPStan extension for PHPUnit is aware of methods like that.

> I'll happily change […]

I'd like to make sure we understand the root cause of why it's not detected before making any change.

Reply:
> I'd like to make sure we understand the root cause of why it's not detected before making any change.

I'll leave it to you then, I'm not familiar enough with phpstan!

Reply:
> As for the code change, I'd rather fail the test explicitly with `$this->fail()`.

Done, FWIW.

Reply:
Scrutinizer sucks.

> […] the function will implicitly return `null` […]

It is not true. The return will never happen.

Reply:
> I'd like to make sure we understand the root cause of why it's not detected before making any change.

@morozov Are you going for the root cause in this PR or will you create a separate one?

Reply:
I'd like to address the root cause. Apparently, PHPStan ignores this issue because of:
https://github.com/doctrine/dbal/blob/f9b3e7908f548a96d4ec62a30c8cb9fed7f95c7f/phpstan.neon.dist#L38-L39

Instead of being applied to a specific class method, it's applied to the entire codebase. Removing this line makes PHPStan find the issue as well as 43 others:
```
$ phpstan a tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php

 ------ ------------------------------------------------------------------------------------- 
  Line   SqlitePlatformTest.php                                                               
 ------ ------------------------------------------------------------------------------------- 
  274    Method Doctrine\Tests\DBAL\Platforms\SqlitePlatformTest::getGenerateForeignKeySql()  
         should return string but returns null.                                               
 ------ ------------------------------------------------------------------------------------- 
```
@BenMorel please try to rework the PHPStan configuration so that it's only applied to where it's supposed to and try to fix the rest of the issues.

Reply:
The problem is probably because the `|` in the `string|null` type is not escaped, so it's interpreted by the regex engine instead of being used as a literal. We should escape all occurrences of `|` where it's part of the expected error message like it's done it:
https://github.com/doctrine/dbal/blob/f9b3e7908f548a96d4ec62a30c8cb9fed7f95c7f/phpstan.neon.dist#L16-L17

Reply:
First of all, when I run `phpstan analyze` locally, I get several times this warning:

> Access to undefined constant PDO::PGSQL_ATTR_DISABLE_PREPARES

How do I get rid of it? Should I report a bug with phpstan?

Reply:
> The problem is probably because the | in the string|null type is not escaped, so it's interpreted by the regex engine instead of being used as a literal.

Good catch! Fixed this in 8f286d31c5b48d58ae71d47a02e3d4fafd279ef5, and now phpstan reports 42 errors. Working on fixing them.

Reply:
> Should I report a bug with phpstan?

Just for a future reference: this is a current limitation of PHPStan and the design flaw of PDO. Some PDO extensions define constants on the already existing `PDO` class. If `pdo` is loaded but `pdo_pgsql` isn't, PHPStan cannot add the missing constants from the `pdo_pgsql` stubs because it uses runtime reflection instead of static.